### PR TITLE
Allow user to configure location of the Google Cloud SDK

### LINKED ISSUE #231: App Engine Standard (Custom) runtime should autolocate the cloud SDK
image

using PathResolver from app-tools-lib-for-java

### PR DESCRIPTION
**Pull request by briandealwis**
Add support for the user to configure the location of the Google Cloud SDK:
  - adds a preference page and named preference (location of the pref page probably needs tweaking in light of the new `Google Cloud Tools` preference page in #243)
  - adds a static accessor `com.google.cloud.tools.eclipse.sdk.internal.CloudSdkProvider` to retrieve and setup a `com.google.cloud.tools.appengine.cloudsdk.CloudSdk` instance or location
  - adds support for injecting a `CloudSdk` instance using E4's DI (useful for `ViewPart` and other workbench objects)
  - cuts over most uses of `com.google.cloud.tools.app.lib`/`app-tools-lib-for-java`'s `PathResolver` to `CloudSdkProvider`

Fixes #231

### CODE REVIEW

**Comment by briandealwis** [comment]
Fixed #231

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -6,15 +6,16 @@ Bundle-Version: 0.1.0.qualifier
...
+Require-Bundle: com.google.cloud.tools.eclipse.appengine.ui,
  javax.servlet;bundle-version="2.5.0",
  javax.servlet.jsp;bundle-version="2.2.0",
  org.eclipse.core.resources,
  org.eclipse.equinox.common,
  org.eclipse.ui,
  org.eclipse.ui.ide,
- org.eclipse.jst.j2ee.web
+ org.eclipse.jst.j2ee.web,
+ com.google.cloud.tools.app.lib;bundle-version="0.1.0",
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/META-INF/MANIFEST.MF]: we still need this one here?

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java [hunk 2]
```diff
@@ -153,7 +155,11 @@ static void installAppEngineRuntime(IFacetedProject project, IProgressMonitor mo
   }
 
   private static String findCloudSdk() {
-    return PathResolver.INSTANCE.getCloudSdkPath().toString();
+    File location = CloudSdkProvider.getDefaultSdkLocation();
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java]: We don't want the default location. We want the actual location. If that's what the method provides, then we should rename it to just getLocation.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java [hunk 3]
```diff
@@ -153,7 +155,11 @@ static void installAppEngineRuntime(IFacetedProject project, IProgressMonitor mo
...
+    if (location == null) {
+      throw new NullPointerException("Could not find Google Cloud SDK");
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java]: Do we want to do this here? What happens to the user if it isn't found?

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferencePage.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import java.io.File;
+
+public class CloudSdkPreferencePage extends FieldEditorPreferencePage
+    implements IWorkbenchPreferencePage {
+  private IWorkbench workbench;
+  private DirectoryFieldEditor sdkLocation;
+
+  public CloudSdkPreferencePage() {
+    super(GRID);
+    // setPreferenceStore(activator.getDefault().getPreferenceStore());
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferencePage.java]: delete

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferencePage.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    addField(sdkLocation);
+  }
+
+  protected boolean validateSdk(File location) {
+    try {
+      new CloudSdk.Builder().sdkPath(location).build().validate();
+    } catch (AppEngineException e) {
+      // accept a seemingly invalid location in case the SDK organization
+      // has changed and this code is out of date
+      System.out.println("No SDK found: " + e);
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferencePage.java]: remove

File: plugins/com.google.cloud.tools.eclipse.sdk/META-INF/MANIFEST.MF [hunk 6]
```diff
@@ -0,0 +1,20 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: Support for Google Cloud SDK within the Eclipse IDE
+Bundle-SymbolicName: com.google.cloud.tools.eclipse.sdk;singleton:=true
+Bundle-Version: 0.1.0.qualifier
+Bundle-Vendor: Google, Inc.
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Bundle-ActivationPolicy: lazy
+C: Bundle-Localization: plugin
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/META-INF/MANIFEST.MF]: what is this C: here?

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import org.eclipse.e4.core.di.IInjector;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.WeakHashMap;
+
+/**
+ * Provides an {@link CloudSdk} instance suitable for injection using the E4 DI mechanisms. As the
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: nit: an --> a
DI --> dependency injection
what is E4?

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+   * A list of IEclipseContexts that have been referenced that need to be updated on preference
+   * change
+   */
+  private static final Map<IEclipseContext, IEclipseContext> referencedContexts =
+      Collections.synchronizedMap(new WeakHashMap<IEclipseContext, IEclipseContext>());
+
+  /** Cloud SDK location has been changed: trigger any necessary updates */
+  public static void sdkPathChanged(Object newPath) {
+    List<IEclipseContext> contexts;
+    synchronized (referencedContexts) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: Why synchronized on referenceContexts here? I'm probably wrong about this here (I am almost always am when it comes to synchronization) but since it's final it shouldn't be mutated. Or is this just about making sure initialization completes before this method starts?

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    for (IEclipseContext c : contexts) {
+      c.set(PreferenceConstants.CLOUDSDK_PATH, newPath);
+    }
+  }
+
+  @Override
+  public Object compute(IEclipseContext context, String contextKey) {
+    Object path = context.get(PreferenceConstants.CLOUDSDK_PATH);
+    referencedContexts.put(context, context);
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: context, context
or 
contextKey, context?

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+  private static final Map<IEclipseContext, IEclipseContext> referencedContexts =
+      Collections.synchronizedMap(new WeakHashMap<IEclipseContext, IEclipseContext>());
+
+  /** Cloud SDK location has been changed: trigger any necessary updates */
+  public static void sdkPathChanged(Object newPath) {
+    List<IEclipseContext> contexts;
+    synchronized (referencedContexts) {
+      contexts = new ArrayList<>(referencedContexts.keySet());
+    }
+    for (IEclipseContext c : contexts) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: c --> context

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+public class CloudSdkContextFunction extends ContextFunction {
+  /**
+   * A list of IEclipseContexts that have been referenced that need to be updated on preference
+   * change
+   */
+  private static final Map<IEclipseContext, IEclipseContext> referencedContexts =
+      Collections.synchronizedMap(new WeakHashMap<IEclipseContext, IEclipseContext>());
+
+  /** Cloud SDK location has been changed: trigger any necessary updates */
+  public static void sdkPathChanged(Object newPath) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: Is Object the most specific type we can use here?

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkProvider.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+ * found in standard locations on the system.
+ */
+public class CloudSdkProvider extends ContextFunction {
+
+  /**
+   * Return the {@link CloudSdk} instance from the configured or discovered Cloud SDK.
+   * 
+   * @return the configured {@link CloudSdk} or {@code null} if no SDK could be found
+   */
+  public static CloudSdk getDefaultSdk() {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkProvider.java]: I think just getSdk.

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkProvider.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    return createBuilder(null).build();
+  }
+
+  /**
+   * Return the location of the configured or discovered Cloud SDK.
+   * 
+   * @return the configured location or {@code null} if the SDK could not be found
+   */
+  public static File getDefaultSdkLocation() {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkProvider.java]: getCloudSdkLocation or maybe better getCloudSdkDirectory.

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkProvider.java [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+  /**
+   * Return a {@link CloudSdk.Builder} instance, suitable for creating a {@link CloudSdk} instance.
+   * {@code path}, if not null, is used as the location to the Cloud SDK, otherwise the configured
+   * or discovered Cloud SDK will be used instead
+   * 
+   * @param path if not {@code null}, overrides the default location for the SDK; can be a
+   *        {@linkplain String}, {@linkplain File}, or {@linkplain Path}.
+   * @return a builder, or {@code null} if the Google Cloud SDK cannot be located
+   */
+  public static CloudSdk.Builder createBuilder(Object path) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkProvider.java]: Can this be more specific than Object?

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkProvider.java [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+   * Attempt to resolve the Google Cloud SDK from the configured or discovered Cloud SDK.
+   * {@code path}, if not {@code null}, is used as the location to the Cloud SDK, otherwise the
+   * configured default will be used instead.
+   * 
+   * @param path if not {@code null}, overrides the default location for the SDK; can be a
+   *        {@linkplain String}, {@linkplain File}, or {@linkplain Path}.
+   * 
+   * @return the location, or {@code null} if not found
+   */
+  private static File resolveSdkLocation(Object path) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkProvider.java]: I think this should return a Path rather than a File, unless Eclipse doesn't deal well with those

**Comment by elharo** [comment]
Is there any way we can test any of this?

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java (hunk 3 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java]: It is called from the method immediately above which uses this NPE approach for communicating errors.  The text shows up in an error dialog.

But that's an excellent point re: prompting the user.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/META-INF/MANIFEST.MF (hunk 1 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/META-INF/MANIFEST.MF]: I'll verify

File: plugins/com.google.cloud.tools.eclipse.sdk/META-INF/MANIFEST.MF (hunk 6 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/META-INF/MANIFEST.MF]: oops: a comment to quiet warnings that there was no plugin.properties file

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java (hunk 7 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: Eclipse 4 is a new layer introduced in the Eclipse Platform that sits above JFace but below the Eclipse 3.x Workbench APIs.

https://wiki.eclipse.org/File:Eclipse_4_Architecture.png

There's some more information on the [Eclipse Wiki](https://wiki.eclipse.org/Eclipse4)

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java (hunk 8 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: The contents of `referencedContexts` could be changed as we iterate over its contents to build the new `ArrayList`.  So we need to get a lock on it first.

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java (hunk 9 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: That's a good point: since it's the keys that are weak, putting the IEC as the value may create a hard link, such that the IEC can never be GCd.  Changed to be (context, contextKey) since WeakHashMap has weak keys.  And I should use Guava's thread safe collections anyways.  That does away with the problem entirely.

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java (hunk 11 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: Should be a String.

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkProvider.java (hunk 14 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkProvider.java]: It's awkward, agreed.  Ok, I'll use `Path`.

Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkProvider.java]: Using `Path` is even more awkward since it has to be continually turned into a `File` to do anything.  Using `File`.

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java (hunk 8 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: What's going to change referencedContexts while we iterate? Most importantly, does whichever thread/object change it also synchronize on it? Otherwise I don't think this helps.

Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: DI can happen on any thread, and a re-injection (as we will do with the IEclipseContext.set() done within the `for` loop) could conceivably cause a fetch of the `CloudSdk` instance.

This goes away with using Guava collections.

**Comment by briandealwis** [comment]
PTAL @elharo.  This introduces a new `CloudSdkPrompter` service that prompts the user if the Google Cloud SDK cannot be found.

You can test the UI prompting by renaming your `google-cloud-sdk` directory and then using _File > New > Google App Engine Standard Java Project_.  You should get an dialog on _Finish_ saying that the Google Cloud SDK cannot be found.  I'll work on UI tests as part of #110.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java [hunk 16]
```diff
@@ -153,7 +152,7 @@ static void installAppEngineRuntime(IFacetedProject project, IProgressMonitor mo
       RuntimeWorkingCopy mutator = (RuntimeWorkingCopy) appEngineRuntimeWorkingCopy;
       Map<String, String> map = mutator.getAttribute(
           "generic_server_instance_properties", new HashMap<>());
-      map.put("cloudSdkDirectory", findCloudSdk());
+      map.put("cloudSdkDirectory", config.getCloudSdkLocation().toString());
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java]: can this ever return null?

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/StandardProjectWizard.java [hunk 17]
```diff
@@ -31,6 +35,13 @@ public void addPages() {
 
   @Override
   public boolean performFinish() {
+    if (config.getCloudSdkLocation() == null) {
+      File location = CloudSdkPrompter.getCloudSdkLocation(new SameShellProvider(getShell()));
+      if (location == null) {
+        return false;
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/StandardProjectWizard.java]: If I read this right, this may leave config.getCloudSdkLocation as null, in which case we need to be ready for this elsewhere.

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferencePage.java [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+    setPreferenceStore(PreferenceInitializer.getPreferenceStore());
+    setDescription("Google Cloud SDK Preferences");
+  }
+
+
+  @Override
+  protected Control createContents(Composite parent) {
+    Composite contents = new Composite(parent, SWT.NONE);
+    Link instructions = new Link(contents, SWT.WRAP);
+    instructions.setText(
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferencePage.java]: These strings should probably be moved out into a messages bundle.

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPrompter.java [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  /**
+   * Prompt the user to install and configure the Google Cloud SDK.
+   * 
+   * @param shellProvider an object that knows how to obtain a shell; may be {@code null}
+   * @return true if the user appears to have configured the SDK, or false if the SDK is unavailable
+   */
+  static boolean promptForSDK(IShellProvider shellProvider) {
+    if (!MessageDialog.openQuestion(null, "Google Cloud SDK Not Configured",
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPrompter.java]: message bundle?

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/CloudSdkProvider.java [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+
+import org.eclipse.e4.core.contexts.ContextFunction;
+import org.eclipse.jface.preference.IPreferenceStore;
+
+import java.io.File;
+import java.nio.file.Path;
+
+/**
+ * Utililty to find the Google Cloud SDK either at locations configured by the user or otherwise
+ * found in standard locations on the system.
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/CloudSdkProvider.java]: delete "otherwise found"

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/Activator.java [hunk 21]
```diff
@@ -0,0 +1,32 @@
+
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/Activator.java]: delete blank line

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+
+  private static File toFile(Object path) {
+    if (path instanceof File) {
+      return (File) path;
+    } else if (path instanceof Path) {
+      return ((Path) path).toFile();
+    } else if (path instanceof String) {
+      return new File((String) path);
+    }
+    return null;
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: Maybe we should just throw an exception here instead? At the very least would be easier to debug and diagnose than a NullPointerException higher up the stack.

**Comment by elharo** [comment]
When I try to run this, it seems to get stuck on validating the SDK. I'm not sure if I've got this set up right. Perhaps merge with master and repush?

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java (hunk 16 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java]: Silly error on my part

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferencePage.java (hunk 18 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferencePage.java]: Ok.

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferencePage.java [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+        "Google Cloud Tools for Eclipse requires the <a href=\"https://cloud.google.com/sdk/\">Google Cloud SDK</a>. "
+            + "Please <a href=\"https://cloud.google.com/sdk/\">install the SDK</a> "
+            + "and configure its location below.");
+    instructions.addSelectionListener(new SelectionAdapter() {
+      @Override
+      public void widgetSelected(SelectionEvent e) {
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferencePage.java]: Oops; I'll enable Checkstyle and fix those.

File: plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java (hunk 16 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.newproject/src/com/google/cloud/tools/eclipse/appengine/newproject/CreateAppEngineStandardWtpProject.java]: You need to merge with master. I think this part of the code has been removed (and if it hasn't it should be).

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java (hunk 22 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: `null` is a valid argument.  Will log if a non-null, non-file'ish object is provided.

**Comment by briandealwis** [comment]
> When I try to run this, it seems to get stuck on validating the SDK. I'm not sure if I've got this set up right. Perhaps merge with master and repush?

I think the warning message wasn't being removed.  Fixed that.  PTAL.

File: plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.Collections;
+import java.util.Set;
+import java.util.logging.Logger;
+
+/**
+ * Provides a {@link CloudSdk} instance suitable for injection using the E4 dependency injection
+ * mechanisms. As the configured location may change, we fetch the path from the context to ensure
+ * that we recompute the CloudSdk instance on path change.
+ */
+public class CloudSdkContextFunction extends ContextFunction {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunction.java]: This class shouldn't be too hard to unit test, at least in part.

**Comment by elharo** [comment]
Before I run this again, can you try merging it to master?

**Comment by briandealwis** [comment]
PTAL @elharo

File: plugins/com.google.cloud.tools.eclipse.sdk.test/META-INF/MANIFEST.MF [hunk 25]
```diff
@@ -1,7 +1,7 @@
 Manifest-Version: 1.0
 Bundle-ManifestVersion: 2
 Bundle-Name: Tests for Google Cloud SDK within the Eclipse IDE
-Bundle-SymbolicName: com.google.cloud.tools.eclipse.sdk.test;singleton:=true
+Bundle-SymbolicName: com.google.cloud.tools.eclipse.sdk.internal;singleton:=true
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk.test/META-INF/MANIFEST.MF]: why this change, and without changing the directory name?

File: plugins/com.google.cloud.tools.eclipse.sdk.test/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunctionTest.java [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+
+import java.nio.file.Path;
+
+public class CloudSdkContextFunctionTest {
+  IEclipseContext context;
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk.test/src/com/google/cloud/tools/eclipse/sdk/internal/CloudSdkContextFunctionTest.java]: private

File: plugins/com.google.cloud.tools.eclipse.sdk.test/src/com/google/cloud/tools/eclipse/sdk/internal/ReflectionUtil.java [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.eclipse.sdk.internal;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+public class ReflectionUtil {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk.test/src/com/google/cloud/tools/eclipse/sdk/internal/ReflectionUtil.java]: Blech. I wish eclipse were more testable

File: plugins/com.google.cloud.tools.eclipse.sdk.test/src/com/google/cloud/tools/eclipse/sdk/internal/ReflectionUtil.java [hunk 28]
```diff
@@ -0,0 +1,49 @@
...
+      throws IllegalArgumentException, IllegalAccessException, NoSuchFieldException,
+      SecurityException {
+    Field field = object.getClass().getDeclaredField(fieldName);
+    field.setAccessible(true);
+    return fieldType.cast(field.get(object));
+  }
+
+  /** Invoke a method. */
+  public static <T> T invoke(Object object, String methodName, Class<T> returnType,
+      Object... paramters)
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk.test/src/com/google/cloud/tools/eclipse/sdk/internal/ReflectionUtil.java]: parameters

**Comment by elharo** [comment]
Trying to import this into eclipse I get:

Invalid project description.
OK
/Users/elharo/gcloud-eclipse-tools/plugins/com.google.cloud.tools.eclipse.sdk.test overlaps the location of another project: 'com.google.cloud.tools.eclipse.sdk.test'

**Comment by elharo** [comment]
builds for me on linuix but not on mac (CLI)

**Comment by chanseokoh** [comment]
In #248, I created the new bundle 'c.g.c.t.eclipse.preferences' in #248, thinking initially that this bundle will accommodate all preference pages and settings across all bundles in a single place.

However, it looks like @briandealwis is taking a more fine-grained and modularized model for preference pages. If that's the case, I think it makes more sense to rename 'c.g.c.eclipse.preferences' to reflect that it is only usagetracker-specific, or I could even merge the bundle into 'c.g.c.t.eclipse.usagetracker'.

In any case, we need to figure out later how we will organize our current and future preference pages.

File: plugins/com.google.cloud.tools.eclipse.sdk.test/src/com/google/cloud/tools/eclipse/sdk/internal/ReflectionUtil.java (hunk 27 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk.test/src/com/google/cloud/tools/eclipse/sdk/internal/ReflectionUtil.java]: These are to access fields and methods on `CloudSdk` ;)

**Comment by elharo** [comment]
@briandealwis Do you want to merge this now, and then  fix the preference pages or fix the preference pages first? I'm not sure what the preference pane should ultimately look like. E.g. is it one pane or several categories under Google Cloud Tools?

**Comment by briandealwis** [comment]
@chanseokoh Thanks for bringing this up, and I wondered about this too.  From the user's perspective it can be overwhelming to have different pages.  We don't have a facility in Eclipse for contributing to a single pref page, but it's easy enough to set up a mini framework.  I'll open a new issue to address this.

**Comment by briandealwis** [comment]
@elharo:
> /Users/elharo/gcloud-eclipse-tools/plugins/com.google.cloud.tools.eclipse.sdk.test overlaps the location of another project: 'com.google.cloud.tools.eclipse.sdk.test'

Sounds like you already have it imported.
     
> builds for me on linuix but not on mac (CLI)

Weird as it build fine for me.  What's the error?

```
[INFO] com.google.cloud.tools.eclipse.usagetracker ........ SUCCESS [  1.129 s]
[INFO] com.google.cloud.tools.eclipse.usagetracker.test ... SUCCESS [  0.046 s]
[INFO] gcp-eclipse-mars ................................... SUCCESS [  0.004 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 29.704 s
[INFO] Finished at: 2016-06-24T13:15:30-04:00
[INFO] Final Memory: 87M/328M
[INFO] ————————————————————————————————————
```

File: plugins/com.google.cloud.tools.eclipse.sdk.test/META-INF/MANIFEST.MF (hunk 25 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk.test/META-INF/MANIFEST.MF]: That was an errant package rename :-(

**Comment by briandealwis** [comment]
@elharo I think [that `.sdk.internal` name change for `.sdk.test`](https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/pull/255#discussion-diff-68308877) probably accounts for the import error.  Fixed those and the above.

**Comment by briandealwis** [comment]
@elharo I think we should merge this now and coalesce later.  The usage tracking and SDK location would seem appropriate to be on the same Google Cloud Tools page at the top level.

**Comment by elharo** [comment]
OK, LGTM.
